enhancements/maintenance/machine-maintenance-operator.md

```diff
@@ -0,0 +1,237 @@
+
+A POC of the idea can be found [here](https://github.com/dofinn/machine-maintenance-operator).
+
+## Motivation
+
+Currently the machine-api marks a machine as stopped when it has been terminated/stopped by the cloud provider. A provider may terminate/stop an instance when there is an associated maintenance scheduled for it. This is also the case for users manually terminating/stopping machines via the console. The MMO is a proactive approach to executing the required action (for example, delete target machine CR) to enable the machine-api to manage machinesets. This operator would work in conjunction with MachineHealthCheck implementation. 
+
+## Requirements
+That the machine-api collects scheduled maintenances from cloud providers and posts them in the status of each machine's CR
+
+### Goals
+
+List the specific goals of the proposal. How will we know that this has succeeded?
+The machine-maintenance-operator will:
+* detect maintenances from machine CRs and execute them manually at their earliest and most convenient time
+
+### Non-Goals
+
+This is not a solution for managing maintenances or machine state for master machine roles. 
+
+## Proposal
+
+### User Stories [optional]
+
+#### Story 1
+As a customer using OKD/OCP/OSD, I want my cluster/s to be proactive in handling events that are initiated by the cloud provider hosting my cluster. This will assure me that OKD/OCP/OSD accounts for machines in the cluster that require terminating or stopping to be proactively maintained and addressed with intention. 
+
+### Implementation Details/Notes/Constraints [optional]
+
+Constraints: 
+* This implementation will require the machine-api to query cloud providers for scheduled maintenances and publish them in the machine's CR. 
+* GCP only allows maintenances to be queried from the node itself -> `curl http://metadata.google.internal/computeMetadata/v1/instance/maintenance-event -H "Metadata-Flavor: Google"`
+
+This operator will iterate through the machineList{} and inspect each machine CR for scheduled maintenances. If a maintenance is found, the controller will validate the state of the cluster prior to performing any maintenance. For example; is the cluster upgrading? is the cluster already performing a maintenance?
+
+These processes will only hold true for infra and worker roles of machines within the cluster. 
+
+If a scheduled maintenance is detected for a master node, an alert should be raised.
+
+Not every cloud provider has the same kind of maintenances that inherit that same degradation on the cluster.  Some of the features may need to be toggleable per provider. 
+
+Example: (AWS maintenance)[https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/monitoring-instances-status-check_sched.html#types-of-scheduled-events] vs (GCP maintenance)[https://cloud.google.com/compute/docs/storing-retrieving-metadata#maintenanceevents].
```

Comment:
Please explain in the enhancement text the points that readers are expected to take from this comparison.

Reply:
Thanks. Resolved: c760ac9aad824b98bf380e55d9855d8c6a092e1b